base-env.lua

```diff
@@ -236,52 +250,67 @@ local function tuple_of_impl(syntax, env)
-			exprs.inferred_expression(utils.accept_with_env, env)
-		),
-	}, metalang.failure_handler, nil)
-	if not ok then
-		return ok, name
-	end
-	return true, name, type_env.val, type_env.env, tail
-end, "pure_ascribed_name_with_tail")
-
-local ascribed_name_with_tail = metalang.reducer(function(syntax, env, prev, names)
-	-- print("ascribed_name trying")
-	-- p(syntax)
-	-- print(prev:pretty_print())
-	-- print("is env an environment? (start of ascribed name)")
-	-- print(env.get)
-	-- print(env.enter_block)
-	---@cast env Environment
-	local shadowed, env = env:enter_block()
-	local anchor = syntax.anchor
-	if not anchor then
-		anchor = "bug in ascribed_name_with_tail"
-	end
-	env = env:bind_local(terms.binding.annotated_lambda("#prev", prev, anchor, terms.visibility.explicit))
-	local ok, prev_binding = env:get("#prev")
-	if not ok then
-		error "#prev should always be bound, was just added"
-	end
-	env = env:bind_local(terms.binding.tuple_elim(names, prev_binding))
-	local ok, name, val, env, tail =
-		syntax:match({ pure_ascribed_name_with_tail(metalang.accept_handler, env) }, metalang.failure_handler, nil)
-	if not ok then
-		return ok, name
-	end
-	---@type Environment
-	local env, val = env:exit_block(val, shadowed)
-	-- print("is env an environment? (end of ascribed name)")
-	-- print(env.get)
-	-- print(env.enter_block)
-	return true, name, val, env, tail
-end, "ascribed_name_with_tail")
+local pure_ascribed_name_with_tail = metalang.reducer(
+	---@type lua_operative
+	function(syntax, env)
+		local ok, name, type_env, tail = syntax:match({
+			metalang.listtail(
+				metalang.accept_handler,
+				metalang.issymbol(metalang.accept_handler),
+				metalang.symbol_exact(metalang.accept_handler, ":"),
+				exprs.inferred_expression(utils.accept_with_env, env)
+			),
+		}, metalang.failure_handler, nil)
+		if not ok then
+			return ok, name
+		end
+		return true, name, type_env.val, type_env.env, tail
+	end,
+	"pure_ascribed_name_with_tail"
+)
+
+local ascribed_name_with_tail = metalang.reducer(
+	---@param syntax ConstructedSyntax
+	---@param env Environment
+	---@param prev inferrable
+	---@param names string[]
+	---@return boolean
+	---@return string
+	---@return unknown
+	---@return Environment?
+	---@return unknown
+	function(syntax, env, prev, names)
+		-- print("ascribed_name trying")
+		-- p(syntax)
+		-- print(prev:pretty_print())
+		-- print("is env an environment? (start of ascribed name)")
+		-- print(env.get)
+		-- print(env.enter_block)
+		local shadowed, env = env:enter_block()
+		local anchor = syntax.anchor
+		if not anchor then
+			anchor = "bug in ascribed_name_with_tail"
+		end
+		env = env:bind_local(terms.binding.annotated_lambda("#prev", prev, anchor, terms.visibility.explicit))
+		local ok, prev_binding = env:get("#prev")
+		if not ok then
+			error "#prev should always be bound, was just added"
+		end
+		env = env:bind_local(terms.binding.tuple_elim(names, prev_binding))
```

Comment:
Once again, prev_binding comes from the environment and is thus a binding, but is passed into tuple_elim which the generated types claim only accepts an inferrable. Either the generated types are wrong, or this is wrong.